app/post.go

```diff
@@ -89,7 +97,74 @@ func (a *App) CreatePostMissingChannel(post *model.Post, triggerWebhooks bool) (
 	return a.CreatePost(post, channel, triggerWebhooks)
 }
 
-func (a *App) CreatePost(post *model.Post, channel *model.Channel, triggerWebhooks bool) (*model.Post, *model.AppError) {
+// deduplicateCreatePost attempts to make posting idempotent within a caching window.
+func (a *App) deduplicateCreatePost(post *model.Post) (foundPost *model.Post, err *model.AppError) {
+	// We rely on the client sending the pending post id across "duplicate" requests. If there
+	// isn't one, we can't deduplicate, so allow creation normally.
+	if post.PendingPostId == "" {
+		return nil, nil
+	}
+
+	const unknownPostId = ""
+
+	for attempt := 1; attempt <= 3; attempt++ {
```

Comment:
I'm not a general fan of repeatedly trying something in the hope that it works, but I really wanted to avoid client changes to handle a "duplicate post". With these changes, unless posting is most likely broken, the client will just see the same payload as if they had been the ones to create the post in the first place.